### PR TITLE
feat(scoop-update): Stash uncommitted changes before update

### PR DESCRIPTION
**Pull request by niheaven**
#### Description

Auto stash uncommitted changes of scoop before update.

Now the action is done in scoop core, should this be applied to buckets too?

#### Motivation and Context

- Relates to https://github.com/ScoopInstaller/Scoop/pull/5089#issuecomment-1207388756

#### How Has This Been Tested?

image

And this could be tested by checking out this PR, adding some dummy files in scoop dir and running `scoop update`.

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have ensured that I am targeting the `develop` branch.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.
- [x] I have added an entry in the CHANGELOG.

### CODE REVIEW

**Comment by rashil2000** [comment]
I think you missed the `-C $currentdir` flag.

But on a second note, stashing changes is an unexpected behaviour for users. I think it is just better to fail the update (and Git automatically shows the reason why the update failed). It is important to let users know they have uncommitted changes, fail gracefully, and then let the user decide whether to stash the changes themselves.

> Now the action is done in scoop core, should this be applied to buckets too?

Bucket update sometimes fail due to EOL conflicts, this change will create multiple stashes there and somewhat pollute the working tree.

**Comment by niheaven** [comment]
I'll add a config option to control it: stash, prompt, cancel.

**Comment by niheaven** [comment]
image

Done.

File: libexec/scoop-update.ps1 [hunk 1]
```diff
@@ -98,6 +98,16 @@ function update_scoop() {
         $isRepoChanged = !($currentRepo -match $configRepo)
         $isBranchChanged = !($currentBranch -match "\*\s+$configBranch")
 
+        # Stash uncommitted changes
+        if (git -C "$currentdir" diff HEAD --name-only) {
+            if (get_config autostash_on_conflict) {
+                warn "Uncommitted changes detected. Stashing..."
+                git -C "$currentdir" stash push -m "WIP at $([System.DateTime]::Now.ToString('o'))" -u -q
+            } else {
+                abort "Uncommitted changes detected. Updating Aborted."
```
Comment (rashil2000) [inline libexec/scoop-update.ps1:107]:
```suggestion
                warn "Uncommitted changes detected. Update aborted."
                return
```

`abort` will exit the entire script, which we don't want

Comment (niheaven) [reply libexec/scoop-update.ps1:107]: Message from Copilot lol.

**Comment by niheaven** [comment]
Any other comments? This should be merged now.

**Review by rashil2000 — DISMISSED** [review]

**Pull request merged**